tests/unit/v1/async/test_async_transaction.py

```diff
@@ -0,0 +1,1055 @@
+        self.assertIsNone(transaction._id)
+
+    def test__add_write_pbs_failure(self):
+        from google.cloud.firestore_v1.base_transaction import _WRITE_READ_ONLY
+
+        batch = self._make_one(mock.sentinel.client, read_only=True)
+        self.assertEqual(batch._write_pbs, [])
+        with self.assertRaises(ValueError) as exc_info:
+            batch._add_write_pbs([mock.sentinel.write])
+
+        self.assertEqual(exc_info.exception.args, (_WRITE_READ_ONLY,))
+        self.assertEqual(batch._write_pbs, [])
+
+    def test__add_write_pbs(self):
+        batch = self._make_one(mock.sentinel.client)
+        self.assertEqual(batch._write_pbs, [])
+        batch._add_write_pbs([mock.sentinel.write])
+        self.assertEqual(batch._write_pbs, [mock.sentinel.write])
+
+    def test__clean_up(self):
+        transaction = self._make_one(mock.sentinel.client)
+        transaction._write_pbs.extend(
+            [mock.sentinel.write_pb1, mock.sentinel.write_pb2]
+        )
+        transaction._id = b"not-this-time-my-friend"
+
+        ret_val = transaction._clean_up()
+        self.assertIsNone(ret_val)
+
+        self.assertEqual(transaction._write_pbs, [])
+        self.assertIsNone(transaction._id)
+
+    @pytest.mark.asyncio
+    async def test__begin(self):
+        from google.cloud.firestore_v1.services.firestore import (
+            client as firestore_client,
+        )
+        from google.cloud.firestore_v1.types import firestore
+
+        # Create a minimal fake GAPIC with a dummy result.
+        firestore_api = mock.create_autospec(
+            firestore_client.FirestoreClient, instance=True
+        )
+        txn_id = b"to-begin"
+        response = firestore.BeginTransactionResponse(transaction=txn_id)
+        firestore_api.begin_transaction.return_value = response
+
+        # Attach the fake GAPIC to a real client.
+        client = _make_client()
+        client._firestore_api_internal = firestore_api
+
+        # Actually make a transaction and ``begin()`` it.
+        transaction = self._make_one(client)
+        self.assertIsNone(transaction._id)
+
+        ret_val = await transaction._begin()
+        self.assertIsNone(ret_val)
+        self.assertEqual(transaction._id, txn_id)
+
+        # Verify the called mock.
+        firestore_api.begin_transaction.assert_called_once_with(
+            request={"database": client._database_string, "options": None},
+            metadata=client._rpc_metadata,
+        )
+
+    @pytest.mark.asyncio
+    async def test__begin_failure(self):
+        from google.cloud.firestore_v1.base_transaction import _CANT_BEGIN
+
+        client = _make_client()
+        transaction = self._make_one(client)
+        transaction._id = b"not-none"
+
+        with self.assertRaises(ValueError) as exc_info:
+            await transaction._begin()
+
+        err_msg = _CANT_BEGIN.format(transaction._id)
+        self.assertEqual(exc_info.exception.args, (err_msg,))
```

Comment:
This feels like a hollow assertion since we're constructing the expected result in the exact same why the internal code is. It may be better to actually write out the expected message to verify the internal creation of it does what is desired.

Reply:
That's a good point. If the construction of the error message was marginally more complex I would agree with you hands down, but it feels as though this is acceptable as it is very close to comparing a defined constant string. @crwilcox do you have any strong opinions about this?

Reply:
A lot of the exception checking is fragile. We already encountered an issue with this in the microgen conversion. Basically we can use the `assertRaisesRegex` instead but it isn't much better. We could mock the var so we could see it was called in the impl properly? But again not sure it is better?

TL;DR - ack